### PR TITLE
feat: sdk 1800 migration code refactor

### PR DESCRIPTION
**Pull request by itsdebs**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] Version upgraded (project, README, gradle, podspec etc)
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: android/src/main/java/com/rudderstack/android/internal/plugins/PlatformInputsPlugin.kt [hunk 1]
```diff
@@ -86,6 +97,20 @@ internal class PlatformInputsPlugin : Plugin, LifecycleListenerPlugin {
...
+    override fun updateConfiguration(configuration: Configuration) {
+        if (configuration !is ConfigurationAndroid) return
+        if (autoCollectAdvertisingId != configuration.autoCollectAdvertId)
+            autoCollectAdvertisingId = configuration.autoCollectAdvertId
+        synchronized(this) {
+            if (_advertisingId != configuration.advertisingId)
+                _advertisingId = configuration.advertisingId
+        }
+        collectDeviceId = configuration.collectDeviceId
     }
```
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply android/src/main/java/com/rudderstack/android/internal/plugins/PlatformInputsPlugin.kt:114]: I discussed this with @itsdebs offline and he helped me in understanding that `updateConfiguration` method is required in cases where config gets updated after the SDK gets initialised e.g., using `putAnonymousId` API.

Also, as of now we are accepting values like `advertisingId` and `anonymousId` through configuration object (while SDK init). But we are going to discuss if this behaviour is really needed. 
If it's not needed then we need to look forward in removing the `updateConfiguration` method from here.

File: android/src/main/java/com/rudderstack/android/internal/infrastructure/LifecycleObserverPlugin.kt [hunk 2]
```diff
@@ -22,15 +22,23 @@ import com.rudderstack.core.Analytics
...
 
+/**
+ * We use a [Plugin] instead of [InfrastructurePlugin] as sending events from Infrastructure plugin might
+ * not ensure all plugins to be ready
+ *
+ * @property currentMillisGenerator
+ */
 internal class LifecycleObserverPlugin(val currentMillisGenerator: (() -> Long) = { SystemClock.uptimeMillis() }) :
-    InfrastructurePlugin, LifecycleListenerPlugin {
+    Plugin, LifecycleListenerPlugin {
```
Comment (1abhishekpandey) [inline android/src/main/java/com/rudderstack/android/internal/infrastructure/LifecycleObserverPlugin.kt:41]: This Interface change has broken the `LifecycleObserverPluginTest` test cases and we are getting following errors:
> Unresolved reference: shutdown.

Please do the required refactoring!

File: android/src/main/java/com/rudderstack/android/internal/infrastructure/ReinstatePlugin.kt [hunk 3]
```diff
@@ -69,84 +61,80 @@ internal class ReinstatePlugin : InfrastructurePlugin {
...
+        )
+        _analytics?.resetV1SessionValues()
+        _analytics?.migrateV1Build()
+        _analytics?.migrateV1Version()
+        if (_isDbBeingMigrated.compareAndSet(false, true)) {
+            _analytics?.androidStorage?.migrateV1StorageToV2 {
+                setReinstated(true)
+            }
+        }
+        _analytics?.androidStorage?.deleteV1SharedPreferencesFile()
```
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply android/src/main/java/com/rudderstack/android/internal/infrastructure/ReinstatePlugin.kt:137]: We have decided to keep this method for now.

File: android/src/main/java/com/rudderstack/android/internal/RudderPreferenceManager.kt [hunk 4]
```diff
@@ -180,16 +192,46 @@ internal class RudderPreferenceManager(application: Application,
...
+    fun resetV1LastActiveTimestamp() {
+        preferencesV1.edit().remove(RUDDER_SESSION_LAST_ACTIVE_TIMESTAMP_KEY).apply()
+    }
+    fun resetV1VersionName(){
+        preferencesV1.edit().remove(RUDDER_APPLICATION_VERSION_KEY).apply()
+    }
+
+    fun resetV1Build() {
+        preferencesV1.edit().remove(RUDDER_APPLICATION_BUILD_KEY).apply()
+    }
```
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply android/src/main/java/com/rudderstack/android/internal/RudderPreferenceManager.kt:226]: Over the call with @itsdebs, we have decided to keep all the reset methods as currently it is not possible to delete the v1 shared preference files!

File: android/src/main/java/com/rudderstack/android/internal/plugins/PlatformInputsPlugin.kt [hunk 5]
```diff
@@ -86,6 +97,20 @@ internal class PlatformInputsPlugin : Plugin, LifecycleListenerPlugin {
...
+
+    override fun updateConfiguration(configuration: Configuration) {
+        if (configuration !is ConfigurationAndroid) return
+        if (autoCollectAdvertisingId != configuration.autoCollectAdvertId)
+            autoCollectAdvertisingId = configuration.autoCollectAdvertId
+        synchronized(this) {
+            if (_advertisingId != configuration.advertisingId)
+                _advertisingId = configuration.advertisingId
+        }
+        collectDeviceId = configuration.collectDeviceId
```
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply android/src/main/java/com/rudderstack/android/internal/plugins/PlatformInputsPlugin.kt:113]: The requested changes have already been addressed in this incremental PR: https://github.com/rudderlabs/rudder-sdk-android/pull/440/files.

File: android/src/main/java/com/rudderstack/android/storage/AndroidStorageImpl.kt [hunk 6]
```diff
@@ -384,9 +396,41 @@ class AndroidStorageImpl(
...
+    override fun migrateV1StorageToV2(callback: (Boolean) -> Unit) {
+        storageExecutor.execute {
+            callback(migrateV1StorageToV2Sync())
+        }
+    }
+
+    override fun deleteV1SharedPreferencesFile() {
+        storageExecutor?.execute {
+            preferenceManager?.deleteV1PreferencesFile()
+        }
```
[the remark this answers is not in the document]
Comment (1abhishekpandey) [reply android/src/main/java/com/rudderstack/android/storage/AndroidStorageImpl.kt:433]: Over the call with @itsdebs, we have decided to handle this deletion of the SharedPreference file after the beta release.

File: android/src/main/java/com/rudderstack/android/internal/infrastructure/ReinstatePlugin.kt [hunk 7]
```diff
@@ -69,84 +61,80 @@ internal class ReinstatePlugin : InfrastructurePlugin {
...
+        _analytics?.migrateContextFromV1()
+        _analytics?.migrateV1AdvertisingId()
+        _analytics?.initializeSessionManagement(
+            _analytics?.androidStorage?.v1SessionId,
+            _analytics?.androidStorage?.v1LastActiveTimestamp
+        )
+        _analytics?.resetV1SessionValues()
+        _analytics?.migrateV1Build()
+        _analytics?.migrateV1Version()
+        if (_isDbBeingMigrated.compareAndSet(false, true)) {
```
[the remark this answers is not in the document]
Comment (itsdebs) [reply android/src/main/java/com/rudderstack/android/internal/infrastructure/ReinstatePlugin.kt:134]: removed this.

File: android/src/main/java/com/rudderstack/android/internal/infrastructure/LifecycleObserverPlugin.kt (hunk 2 above)
Comment (itsdebs) [reply android/src/main/java/com/rudderstack/android/internal/infrastructure/LifecycleObserverPlugin.kt:41]: Not getting this error. Can you recheck? If this exists still, we can get on a call to resolve.

Comment (1abhishekpandey) [reply android/src/main/java/com/rudderstack/android/internal/infrastructure/LifecycleObserverPlugin.kt:41]: This issue doesn't seem to occur now.

File: android/src/main/java/com/rudderstack/android/internal/infrastructure/ReinstatePlugin.kt [hunk 8]
```diff
@@ -69,84 +60,83 @@ internal class ReinstatePlugin : InfrastructurePlugin {
...
-        this.sourceId = sourceId
-        reinstate()
-    }
-
-    override fun updateConfiguration(configuration: Configuration) {
         reinstate()
     }
 
     private fun reinstate() {
-        if (_isReinstated.get()) return
```
Comment (1abhishekpandey) [inline android/src/main/java/com/rudderstack/android/internal/infrastructure/ReinstatePlugin.kt:85]: Are we sure we don't need this flag? As we are setting this flag but not utilising it. If it's not needed can we please consider removing this flag altogether?

**Review by 1abhishekpandey — APPROVED** [review]

**Pull request merged**